fix(UI): replace button with icons in challenge lower jaw

mehdibouaziz opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #47157

<!-- Feel free to add any additional description of changes below this line -->

- replaced the text inside help and reset buttons with icons
- added title and aria-label for tooltip - the tooltip uses the translate function to have a localised text
- removed the btn-block class to allow a more compact UI which will allow more icons in the future

mehdibouaziz committed:
fix: replace button with icons in challenge lower jaw

ahmadabdolsaheb commented:
Solid work here. Here are a few things that need to be addressed before this gets merged.

1) Are the icons your design. If so, please look in client/src/assets/icons for custom svg samples and usage. If not, what is the license?
2) Please add data-cy to the icon buttons and select the reset button accordingly from the related cypress test so the tests pass
3) Could you right align the buttons for consistency with mobile's new learn UI.

alicalimli reviewed (commented):
Nice

mehdibouaziz committed:
fix: update icon buttons to add data-cy, align right, and cleaner conditional display

mehdibouaziz commented:
@ahmadabdolsaheb
I modified accordingly but I could find the proper test in cypress, I don't think the tests for these two buttons were written yet. The previous buttons I replaced were also missing data-cy.
I'm not familiar with writting cypress tests so maybe it needs a new issue ?

Sboonny reviewed (commented):
I can't review it currently, but I am wondering if refactoring the code like below is possible?

nayabatir1 commented:
> Hi @Sboonny, thanks for the remarks! I used {condition ? : : <></>} instead of {condition && } because I read some time ago that the ternary version should be prefered to avoid bugs, after struggling with a bug myself. This [article](https://medium.com/geekculture/stop-using-for-conditional-rendering-in-react-a0f7b96200f8) discusses the issue but suggest to use {condition ? : : null} instead. I'll let you decide which solution you prefer :)
> 
> The missing () around the condition is an error on my part, I'll fix it in the next commit.

Hey @mehdibouaziz, I read the article you shared, there is no harm in using short circuit expression, you just have to make sure that the condition evaluates to boolean and not truthly/falsy.

mehdibouaziz committed:
fix: update cypress test with correct selector, update conditional button syntax

mehdibouaziz commented:
@ahmadabdolsaheb  I found the selector error in the failing cypress test and put the correct one.
@nayabatir1 Hi, thanks for your feedback. True, but I tried with the short circuit and in this case one of the conditions as it was written evaluates to 0 and causes a 0 to render:
![image](https://user-images.githubusercontent.com/31729703/195757845-6a434e0d-4e6b-4dac-aa3a-de7714a53a89.png)
![fcc-conditionnal-0](https://user-images.githubusercontent.com/31729703/195757870-b5940d0b-4676-48c0-8c1c-6405a324f436.JPG)
I switched to ternary with `null` just to be sure :)

nayabatir1 commented:
You can refer to this article https://www.codingem.com/javascript-double-exclamation-operator/

ahmadabdolsaheb commented on client/src/templates/Challenges/classic/lower-jaw.css:
```diff
@@ -0,0 +1,13 @@
+.lower-jaw-icon-bar {
+  overflow: hidden;
+}
+
+.lower-jaw-btn-icon {
+  line-height: 0;
+  margin-right: 2px;
+  float: right;
+}
+
+.lower-jaw-btn-icon:first-child {
+  margin-right: 0;
+}
```
They could live in the editor.css similar to other lower jaw styles.

mehdibouaziz replied:
done: moved to editor.css and increased margin to 10px

ahmadabdolsaheb commented on client/src/templates/Challenges/classic/lower-jaw.tsx:
```diff
@@ -181,32 +182,49 @@ const LowerJaw = ({
-          <hr />
-          <button className='btn-block btn fade-in' onClick={openResetModal}>
-            {t('learn.editor-tabs.restart-step')}
+            <svg
+              fill='currentColor'
+              xmlns='http://www.w3.org/2000/svg'
+              height='24'
+              width='24'
+            >
+              <path d='M12 20q-3.35 0-5.675-2.325Q4 15.35 4 12q0-3.35 2.325-5.675Q8.65 4 12 4q1.725 0 3.3.713 1.575.712 2.7 2.037V4h2v7h-7V9h4.2q-.8-1.4-2.187-2.2Q13.625 6 12 6 9.5 6 7.75 7.75T6 12q0 2.5 1.75 4.25T12 18q1.925 0 3.475-1.1T17.65 14h2.1q-.7 2.65-2.85 4.325Q14.75 20 12 20Z' />
+            </svg>
```
Both custom SVGs should be stand alone react components and placed in client/src/assets/icons.

Feel free to look at other icon component samples.

mehdibouaziz replied:
done for both icons

ahmaxed reviewed (commented):
This for making the modifications. It should be ready to merge after these changes.

mehdibouaziz committed:
fix: add reset and help icon components, move lowerjaw.css to editor.css

mehdibouaziz commented:
All done :)

note that I tried to add parenthesis to the condition for the conditional display of the button, but when committing prettier removes them, they are unnecessary I suppose.
` {isAttemptsLargerThanTest && !challengeIsCompleted ? (`

bbsmooth commented:
Just a minor nitpick. Do we need to float the get help button to the left of the reset button? Usually the tab sequence goes left to right so it seems a little odd to have it go right to left for these two buttons. This isn't technically an accessibility failure though, so I'm not demanding anything here. Although I would highly recommend that we add our own custom focus outline on these buttons (instead of relying on the browser's default, which isn't always so great) so that it is perfectly clear which one has focus. I've been using the outline color `--blue-mid` for things I work on because it has accessible contrast with both our light and dark themes. But I understand that that is not necessarily an official policy.

mehdibouaziz commented:
Hi, thanks for the review!

You mean that with the current solution when using tab the focus is moving from right to left because of float:right ?
I can switch to a flexbox instead if it would improve accessibility, there's already a container bar for the buttons.

Since the help button is hidden at first and appears after a few fails, I thought it would be better to have it appear on the left so it doesn't shuffle around the reset button.

For the focus color I'll let the owners decide, it's easy to add.

bbsmooth commented:
No need to switch to flexbox for accessibility, this is purely about DOM order of element, which flexbox won't change. Although if we eventually add more buttons then I think flexbox would be the better solution here.

Another option might be to place the buttons on the left side instead of the right. Then when we add the help button it would be placed to the right of the reset button and wouldn't cause a shift and the tab order would follow the DOM order.

ahmaxed committed:
feat: adjust icon buttons order

ahmaxed committed:
Merge branch 'main' into fix/replace-button-with-icons-lower-jaw-challenge

ahmadabdolsaheb commented:
Nice catch @bbsmooth, I just simplified the styles and the buttons should show up on the left side of the button jaw in the correct DOM order.

@mehdibouaziz, once the tests pass LGMT.

mehdibouaziz committed:
Merge branch 'main' into fix/replace-button-with-icons-lower-jaw-challenge

Sboonny reviewed (changes requested):
nitpick, when someone hovers over the button, the tooltip can block the visual indicator of the other button.

Should we position the tooltip, a little to the top, instead of the bottom of the courser? Or will this affect introduce some sort of CSS bug 🤔 

![image](https://user-images.githubusercontent.com/88248797/196939355-71203d19-a278-4368-9ef6-fceebaaa6ca9.png)

Sboonny reviewed (approved):
#### :sparkles: :tada: **AWESOME!** :tada: :sparkles:

 Hi @mehdibouaziz, thanks for this pull request and for contributing to the code-base for the first time. We are looking forward to more contributions from you in the future.

 Cheers & happy contributing!
